Refactor deprecation warning tests

distantnative opened the pull request:
PHPUnit 10 won't throw errors anymore, which is why we need to switch to a different test setup for our deprecation errors

distantnative committed:
Refactor deprecation warning tests

bastianallgeier commented on tests/Cms/HelpersTestCase.php:
```diff
@@ -0,0 +1,48 @@
+
+		$called = false;
+
+		set_error_handler(
```
Do we need to unset that error handler somehow to avoid side effects?

distantnative replied:
Gets unset in the tearDown

lukasbestle replied:
I think it would be safer to restore it directly in the assertion. Otherwise it will break if multiple assertions are used in one test (as multiple error handlers will be registered but only one will be restored). That would be quite hard to debug.

distantnative replied:
I just want to keep this PR as low key as possible and we have used tearDown already so far for this

lukasbestle replied:
We still need the `hasErrorHandler` check, but IMO that one belongs back into the `HelpersTest` class where the `hasErrorHandler` prop is set by tests.

distantnative replied:
Why? It's needed for `HelpersTestCase::assertError` too. Although I saw that you just removed it, but I don't understand why.

lukasbestle replied:
My commit was for an unrelated method where I spotted it being used superfluously. Doesn't relate to the `assertError()` method.

Until now, the `$hasErrorHandler` prop was only needed for tests that set a custom error handler. If those tests fail, their code stops executing immediately and they cannot reset the handler again, so it has to be done in the `tearDown()` method.

Thinking about it again, the same can happen with the assertion method (didn't realize that before): If the assertion fails, execution stops immediately here as well and we cannot reset the error handler. So we do need the `tearDown()` method and I'm fine with leaving it in the `HelpersTestCase` as it needs to be used there.

However we still have the issue of not being able to call the `assertError()` method multiple times in a single test without accumulating error handlers. We could fix that by turning the `bool $hasErrorHandler = false` prop into an `int $activeErrorHandlers = 0` prop. We can increase that number by one for each error handler we set. In `tearDown()`, we can then reset all handlers in a loop until the count is at 0 again.

distantnative replied:
Do we need to use it more than once per test method?

lukasbestle replied:
Implemented that locally, will commit it with the CI fix.

lukasbestle replied:
> Do we need to use it more than once per test method?

No, but if we ever do, we would cause a hard to debug issue in CI.

lukasbestle commented on tests/Cms/HelpersTestCase.php:
```diff
@@ -0,0 +1,48 @@
+		int $expectedErrorType,
+		string $exptectedErrorMessage,
+		Closure $callback,
+		bool $expectedFailure = true
```
PHPUnit assertions always follow the argument convention `assertSomething($expected, $actual)`. For consistency and to reduce the number of arguments I'd suggest:

```suggestion
		array|false $expected,
		Closure $callback
```

If `$expected` is an array, a failure is expected with `$errorType = $expected[0]` and `$errorMessage = $expected[1]`.

distantnative replied:
There are also standard assertion methods in PHPUnit with only one or more than two arguments. I would rather keep it instead of making it more cryptic.

lukasbestle replied:
You are right that it would become more cryptic. I don't really like either solution, but I can't find one that ticks both the "not cryptic" and "consistent with PHPUnit" boxes. So it's fine for me.

However there's a typo in the second argument name. Could you please fix that?

```diff
- $exptectedErrorMessage
+ $expectedErrorMessage
```

lukasbestle committed:
Handle multiple error handlers per test

We need to pop exactly as many error handlers off the stack as we added, even if e.g. the `assertError()` method was called multiple times

lukasbestle committed:
Fix typo

lukasbestle committed:
Add TODO comment
